### PR TITLE
TST, DOC: Upload devdocs and neps after circleci build

### PR DESCRIPTION
**Pull request by stefanv**
This PR uses CircleCI to deploy the NumPy NEP and development documentation to http://www.numpy.org/neps and http://www.numpy.org/devdocs respectively.  For doing so, we also add a tool that pushes files from a provided directory into a remote repository, overwriting the content there.

The PR also addresses a problem in the masked array documentation, where docstrings are altered dynamically and sometimes introduces multiple `Notes` sections.

### CODE REVIEW

**Comment by stefanv** [comment]
Hrm, looks like numpydoc can't be imported.  Odd.

**Comment by stefanv** [comment]
This is ready for review.

File: numpy/ma/core.py [hunk 1]
```diff
@@ -132,19 +133,17 @@ def doc_note(initialdoc, note):
     if note is None:
         return initialdoc
 
-    # FIXME: disable this function for the moment until we figure out what to
-    # do with it. Currently it may result in duplicate Notes sections or Notes
-    # sections in the wrong place
-    return initialdoc
+    notesplit = re.split('\n\s*?Notes\n\s*?-----', initialdoc)
```
Comment (charris) [inline numpy/ma/core.py]: Need a literal string here for python 3.6+, or else `\\s`.

Comment (charris) [reply numpy/ma/core.py]: You can check for these problems with
```
charris@fc [numpy.git (misc-fixes)]$ python3 tools/find_deprecated_escaped_characters.py numpy/ma/core.py
Errors Found 0
```

**Comment by stefanv** [comment]
Thanks for noticing; that should be fixed now.

**Comment by charris** [comment]
The name `push_to_repo` could be more descriptive, maybe `push_docs_to_repo`?

**Comment by charris** [comment]
Could you add more description of what this PR does up top?

**Comment by stefanv** [comment]
So, the script is fairly generic, it just pushes a directory to a repo.  But happy to rename if that is helpful?  I've updated the PR description.

File: tools/push_to_repo.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+)
+args = parser.parse_args()
+args.dir = os.path.abspath(args.dir)
+
+if not os.path.exists(args.dir):
+    print('Content directory does not exist')
+    sys.exit(1)
+
+
+def run(cmd, stdout=True):
```
Comment (eric-wieser) [inline tools/push_to_repo.py]: What's the purpose of this over `subprocess.check_call`?

File: tools/push_to_repo.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+os.chdir(workdir)
+
+run(['git', 'init'])
+run(['git', 'remote', 'add', 'origin',  args.remote])
+run(['git', 'config', '--local', 'user.name', args.committer])
+run(['git', 'config', '--local', 'user.email', args.email])
+
+print('- committing new content: "%s"' % args.message)
+run(['cp', '-R', os.path.join(args.dir, '.'), '.'])
+run(['git', 'add', '.'], stdout=False)
```
Comment (eric-wieser) [inline tools/push_to_repo.py]: Assuming this is python 3, you can use `subprocess.check_call(..., stdout=subprocess.DEVNULL)`

**Comment by stefanv** [comment]
I've updated the tool according to the feedback.

**Comment by charris** [comment]
Thanks Stefan.

File: tools/ci/push_docs_to_repo.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env python
```
Comment (eric-wieser) [inline tools/ci/push_docs_to_repo.py]: `python3` maybe, to ensure `subprocess.DEVNULL`

File: tools/ci/push_docs_to_repo.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        subprocess.check_call(cmd, stdout=pipe, stderr=pipe)
+    except subprocess.CalledProcessError:
+        print("\n! Error executing: `%s;` aborting" % ' '.join(cmd))
+        sys.exit(1)
+
+
+workdir = tempfile.mkdtemp()
+os.chdir(workdir)
+
+run(['git', 'init'])
```
Comment (eric-wieser) [inline tools/ci/push_docs_to_repo.py]: Is there a reason we create a new history every time?

File: tools/ci/push_docs_to_repo.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    print('Content directory does not exist')
+    sys.exit(1)
+
+
+def run(cmd, stdout=True):
+    pipe = None if stdout else subprocess.DEVNULL
+    try:
+        subprocess.check_call(cmd, stdout=pipe, stderr=pipe)
+    except subprocess.CalledProcessError:
+        print("\n! Error executing: `%s;` aborting" % ' '.join(cmd))
```
Comment (eric-wieser) [inline tools/ci/push_docs_to_repo.py]: Nit: ` ' '.join(shlex.quote(s) for s in cmd)`

File: tools/ci/push_docs_to_repo.py (hunk 5 above)
Comment (njsmith) [reply tools/ci/push_docs_to_repo.py]: It's easier to start over than to correctly make the update in place (you have to worry about things like deleting files that shouldn't exist anymore, etc.). And there's no point in wasting space and time fetching and storing old versions of the rendered HTML; we already have the .rst file history.

**Pull request merged**